components/tidb_query_executors/src/limit_executor.rs

```diff
@@ -32,7 +32,9 @@ impl<Src: BatchExecutor> BatchExecutor for BatchLimitExecutor<Src> {
 
     #[inline]
     fn next_batch(&mut self, scan_rows: usize) -> BatchExecuteResult {
-        let mut result = self.src.next_batch(scan_rows);
+        let mut result = self
+            .src
+            .next_batch(std::cmp::min(scan_rows, self.remaining_rows));
```

Comment:
What if it's something like `select * from (select col, count(*) from table group by col) limit 1`?

Reply:
This SQL will not push down LIMIT to the coprocessor.

Reply:
What if it's something like `select * from table where non_index_col != "" limit 1`?

> ...will not push down...

How to avoid unexpected regression if it's pushed down one day? (For example, all data are within one region)

Reply:
If the condition is never satisfied, then there will be regression, but the regression is only noticeable for AP request.
AP request rarely use LIMIT in statement.

Reply:
condition with limit is very common, perhaps we need another survey? 🤣

Reply:
> condition with limit is very common, perhaps we need another survey? 🤣

Common in TP requests, not in AP. Limit in AP doesn't make much sense because we need statistics for all relevent data.

Reply:
Maybe we can fix regression of selection operator in another issue.

Reply:
> Maybe we can fix regression of selection operator in another issue.

This fix will not affect performance of **selection**. I will post performance test of selection later.